pipenv/utils.py

```diff
@@ -532,12 +551,12 @@ def convert_deps_from_pip(dep):
 
     dependency = {}
 
-    req = [r for r in requirements.parse(dep)][0]
+    req = get_requirement(dep)
```

Comment:
We just modified the path here in the `get_requirement` function but we modify it again below. Is the path adding needed in the function? It looks like we don't use it in either of the two calls this is used.

Reply:
This is the only one I need to push back on -- yes it is needed, and yes it is used.  `get_requirement` takes a string and uses `requirements.parse()` to parse it.  The root of the issue we had with handling relative paths is that `requirements.parse()` can only take either a requirement name (without any path separators, it must be a valid package name according to the pep standards) or a URI.

In the case of a relative path, we need to instantiate a `Requirement` object by converting any existing relative paths to their corresponding `file://` URI.  However, we want to store the relative path in the `Pipfile`, which only happens if `Requirement.path` is set and `Requirement.uri` is not set.  According to the spec, the `file` property is for URIs, but we are writing `path` for the relative path.  So we store the relative path to put in the `req.path` property, but we resolve the `file://` uri for parsing Requirements:

```python
        dep_path = Path(dep)
        if dep_path.exists() and not dep_path.is_absolute():
            path = get_converted_relative_path(dep)
            dep = dep_path.resolve().as_uri()
        ...
    if req.local_file and req.uri and not req.path and path:
        req.path = path
        req.uri = None
    return req
```

The modifications we are doing in `convert_deps_from_pip` don't have any impact to `req.path` in the cases where this code does.  This code can surely be refactored, but the code I have runs only when the user supplies a local relative path (on reflection it should work on absolute paths also) whereas the other code modifies `path` only when there is a file supplied as the `name`, but there is no `path` or `uri`.

I basically left this code alone because there is only one conditional that possibly duplicates anything I've done and I didn't want to refactor the entire function which does quite a bit of heavy lifting, at least not in this PR

Reply:
Hm I reevaluated some of this and in order to fix the `ansible` case above some of this needed to be refactored.  Update incoming.